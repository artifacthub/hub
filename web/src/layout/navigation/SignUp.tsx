import isNull from 'lodash/isNull';
import isUndefined from 'lodash/isUndefined';
import React, { useRef, useState } from 'react';
import { FaEnvelope } from 'react-icons/fa';

import Modal from '../common/Modal';
import CreateAnAccount from './CreateAnAccount';
import OAuth from './OAuth';
import styles from './SignUp.module.css';

interface Loading {
  status: boolean;
  type?: 'log' | 'google' | 'github' | 'oidc';
}

interface Props {
  openSignUp: boolean;
  setOpenSignUp: React.Dispatch<React.SetStateAction<boolean>>;
}

const SignUp = (props: Props) => {
  const form = useRef<HTMLFormElement>(null);
  const [apiError, setApiError] = useState<string | null>(null);
  const [activeSignUp, setActiveSignUp] = useState(false);
  const [successNewAccount, setSuccessNewAccount] = useState(false);
  const [isLoading, setIsLoading] = useState<Loading>({ status: false });

  // Clean API error when form is focused after validation
  const cleanApiError = () => {
    if (!isNull(apiError)) {
      setApiError(null);
    }
  };

  const onCloseModal = () => {
    props.setOpenSignUp(false);
  };

  const submitForm = () => {
    if (form.current) {
      form.current.dispatchEvent(new Event('submit', { cancelable: true }));
    }
  };

  const closeButton = (
    <button className="btn btn-secondary" type="button" disabled={isLoading.status} onClick={submitForm}>
      {!isUndefined(isLoading.type) && isLoading.type === 'log' ? (
        <>
          <span className="spinner-grow spinner-grow-sm" role="status" aria-hidden="true" />
          <span className="ml-2">Signing up...</span>
        </>
      ) : (
        <>Sign up</>
      )}
    </button>
  );

  return (
    <Modal
      header={<div className="h3 m-2 flex-grow-1">Sign up</div>}
      modalClassName={styles.modal}
      open={props.openSignUp}
      onClose={onCloseModal}
      error={apiError}
      cleanError={cleanApiError}
      closeButton={!successNewAccount && activeSignUp ? closeButton : undefined}
    >
      {activeSignUp ? (
        <CreateAnAccount
          ref={form}
          setApiError={setApiError}
          apiError={apiError}
          success={successNewAccount}
          setSuccess={setSuccessNewAccount}
          isLoading={isLoading}
          setIsLoading={setIsLoading}
        />
      ) : (
        <div className="my-auto">
          <div className="h5 mb-5 text-center">Create your account using your email</div>

          <button
            data-testid="signUpBtn"
            type="button"
            onClick={() => setActiveSignUp(true)}
            className="btn btn-outline-secondary btn-block"
            disabled={isLoading.status}
          >
            <div className="d-flex align-items-center">
              <FaEnvelope />
              <div className="flex-grow-1 text-center">Sign up</div>
            </div>
          </button>

          <OAuth separatorClassName="my-5" isLoading={isLoading} setIsLoading={setIsLoading} />
        </div>
      )}
    </Modal>
  );
};

export default SignUp;
