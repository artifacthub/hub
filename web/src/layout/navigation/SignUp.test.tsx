import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';

import SignUp from './SignUp';
jest.mock('../../api');

const setOpenSignUpMock = jest.fn();

const defaultProps = {
  openSignUp: true,
  setOpenSignUp: setOpenSignUpMock,
};

Object.defineProperty(document, 'querySelector', {
  value: () => ({
    getAttribute: () => 'true',
  }),
  writable: true,
});

describe('SignUp', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const result = render(<SignUp {...defaultProps} />);
    expect(result.asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      const { getByText, getByTestId } = render(<SignUp {...defaultProps} />);

      expect(getByText('Create your account using your email')).toBeInTheDocument();
      expect(getByTestId('signUpBtn')).toBeInTheDocument();
      expect(getByText('Github')).toBeInTheDocument();
      expect(getByText('Google')).toBeInTheDocument();
      expect(getByText('OpenID Connect')).toBeInTheDocument();
    });

    it('renders create an account form to click Sign up button', () => {
      const { getByTestId } = render(<SignUp {...defaultProps} />);

      const btn = getByTestId('signUpBtn');
      fireEvent.click(btn);

      waitFor(() => {
        expect(getByTestId('createAnAccountForm')).toBeInTheDocument();
      });
    });
  });
});
