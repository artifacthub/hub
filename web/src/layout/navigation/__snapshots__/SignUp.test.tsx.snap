// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp creates snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="modal-backdrop activeBackdrop"
      data-testid="modalBackdrop"
    />
    <div
      class="modal modal active d-block"
      role="dialog"
    >
      <div
        class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
      >
        <div
          class="modal-content content modal"
        >
          <div
            class="modal-header d-flex flex-row align-items-center header undefined"
          >
            <div
              class="h3 m-2 flex-grow-1"
            >
              Sign up
            </div>
            <button
              class="close"
              data-testid="closeModalBtn"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
            </button>
          </div>
          <div
            class="modal-body p-4 h-100 d-flex flex-column"
          >
            <div
              class="my-auto"
            >
              <div
                class="h5 mb-5 text-center"
              >
                Create your account using your email
              </div>
              <button
                class="btn btn-outline-secondary btn-block"
                data-testid="signUpBtn"
                type="button"
              >
                <div
                  class="d-flex align-items-center"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M502.3 190.8c3.9-3.1 9.7-.2 9.7 4.7V400c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V195.6c0-5 5.7-7.8 9.7-4.7 22.4 17.4 52.1 39.5 154.1 113.6 21.1 15.4 56.7 47.8 92.2 47.6 35.7.3 72-32.8 92.3-47.6 102-74.1 131.6-96.3 154-113.7zM256 320c23.2.4 56.6-29.2 73.4-41.4 132.7-96.3 142.8-104.7 173.4-128.7 5.8-4.5 9.2-11.5 9.2-18.9v-19c0-26.5-21.5-48-48-48H48C21.5 64 0 85.5 0 112v19c0 7.4 3.4 14.3 9.2 18.9 30.6 23.9 40.7 32.4 173.4 128.7 16.8 12.2 50.2 41.8 73.4 41.4z"
                    />
                  </svg>
                  <div
                    class="flex-grow-1 text-center"
                  >
                    Sign up
                  </div>
                </div>
              </button>
              <div
                class="my-5 my-4 position-relative text-center separator"
              >
                <span
                  class="p-3 m-auto separatorContent"
                >
                  or sign in with
                </span>
                <div
                  class="divider"
                />
              </div>
              <div>
                <div
                  class="d-flex flex-column mb-1"
                >
                  <button
                    class="btn btn-outline-secondary mb-3 btn-block btn"
                    type="button"
                  >
                    <div
                      class="d-flex align-items-center"
                    >
                      <img
                        alt="Github Logo"
                        class="logo"
                        src="/static/media/github-mark.svg"
                      />
                      <div
                        class="flex-grow-1 text-center"
                      >
                        Github
                      </div>
                    </div>
                  </button>
                  <button
                    class="btn btn-outline-secondary mb-3 btn-block btn"
                    type="button"
                  >
                    <div
                      class="d-flex align-items-center"
                    >
                      <img
                        alt="Google Logo"
                        class="logo"
                        src="/static/media/google.svg"
                      />
                      <div
                        class="flex-grow-1 text-center"
                      >
                        Google
                      </div>
                    </div>
                  </button>
                  <button
                    class="btn btn-outline-secondary mb-3 btn-block btn"
                    type="button"
                  >
                    <div
                      class="d-flex align-items-center"
                    >
                      <img
                        alt="OpenID Logo"
                        class="logo"
                        src="/static/media/openid.svg"
                      />
                      <div
                        class="flex-grow-1 text-center"
                      >
                        OpenID Connect
                      </div>
                    </div>
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                class="alertWrapper"
                data-testid="alertWrapper"
              />
            </div>
          </div>
          <div
            class="modal-footer p-3"
          >
            <button
              class="btn btn-sm btn-secondary text-uppercase"
              data-testid="closeModalFooterBtn"
              type="button"
            >
              <div
                class="d-flex flex-row align-items-center"
              >
                <svg
                  class="mr-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <div>
                  Close
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
